src/lib.rs

```diff
@@ -8,86 +8,612 @@
+pub enum Source {
+    /// Settings have been retrieved from `pkg-config`
+    PkgConfig,
+    /// Settings have been defined using user defined environnement variables
+    EnvVariables,
+}
+
+#[derive(Debug)]
+/// A system dependency
+pub struct Library {
+    /// From where the library settings have been retrieved
+    pub source: Source,
+    /// libraries the linker should link on
+    pub libs: Vec<String>,
+    /// directories where the compiler should look for libraries
+    pub link_paths: Vec<PathBuf>,
+    /// frameworks the linker should link on
+    pub frameworks: Vec<String>,
+    /// directories where the compiler should look for frameworks
+    pub framework_paths: Vec<PathBuf>,
+    /// directories where the compiler should look for header files
+    pub include_paths: Vec<PathBuf>,
+    /// macros that should be defined by the compiler
+    pub defines: HashMap<String, Option<String>>,
+    /// library version
+    pub version: String,
+}
+
+impl Library {
+    fn from_pkg_config(l: pkg_config::Library) -> Self {
+        Self {
+            source: Source::PkgConfig,
+            libs: l.libs,
+            link_paths: l.link_paths,
+            include_paths: l.include_paths,
+            frameworks: l.frameworks,
+            framework_paths: l.framework_paths,
+            defines: l.defines,
+            version: l.version,
+        }
+    }
+
+    fn from_env_variables() -> Self {
+        Self {
+            source: Source::EnvVariables,
+            libs: Vec::new(),
+            link_paths: Vec::new(),
+            include_paths: Vec::new(),
+            frameworks: Vec::new(),
+            framework_paths: Vec::new(),
+            defines: HashMap::new(),
+            version: String::new(),
+        }
+    }
+
+    /// Create a `Library` by probing `pkg-config` on an internal directory.
+    /// This helper is meant to be used by `Config::add_build_internal` closures
+    /// after having built the lib to return the library information to metadata.
+    ///
+    /// # Arguments
+    ///
+    /// * `pkg_config_dir`: the directory where the library `.pc` file is located
+    /// * `lib`: the name of the library to look for
+    /// * `version`: the minimum version of `lib` required
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut config = metadeps::Config::new();
+    /// config.add_build_internal("mylib", |version| {
+    ///   // Actually build the library here
+    ///   metadeps::Library::from_internal_pkg_config("build-dir",
+    ///       "mylib", version)
+    /// });
+    /// ```
+    pub fn from_internal_pkg_config<P>(
+        pkg_config_dir: P,
+        lib: &str,
+        version: &str,
+    ) -> Result<Self, BuildInternalClosureError>
+    where
+        P: AsRef<Path>,
+    {
+        // save current PKG_CONFIG_PATH so we can restore it
+        let old = env::var("PKG_CONFIG_PATH");
+
+        env::set_var("PKG_CONFIG_PATH", pkg_config_dir.as_ref());
```

Comment:
You might have to include the old path or otherwise some `Require`s of the newly build library might not be found